### PR TITLE
JA4 for TLS and QUIC -- v8

### PR DESCRIPTION
**Pull request by satta**
Previous PR: #10095

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/6379

Changes to previous PR:
- Rebase against current `master`.
- Make JA3/JA4 switchable at compile time using `--disable-ja3`/`--disable-ja4`.
- Add new `feature`s for `requires` keyword: `ja3`,`ja4`.
- Add new feature strings: `HAVE_JA3`, `HAVE_JA4`

```
SV_BRANCH=pr/1634
```

### CODE REVIEW

**Comment by satta** [comment]
Not going to touch the clippy stuff, that's probably already addressed in #10339?

**Comment by catenacyber** [comment]
> Not going to touch the clippy stuff, that's probably already addressed in #10339?

Yes, you can rebase on latest master now

File: doc/userguide/rules/ja4-keywords.rst [hunk 1]
```diff
@@ -0,0 +1,30 @@
+JA4 Keywords
+============
+
+Suricata comes with a JA4 integration (https://github.com/FoxIO-LLC/ja4). JA4,
+as part of the larger JA4+ suite of fingerprints, is used to fingerprint TLS
+clients.
+
+We might in the future consider adding raw and original order strings (JA4_r,
```
Comment (catenacyber) [inline doc/userguide/rules/ja4-keywords.rst:8]: I do not think this should be part of the doc

File: doc/userguide/output/eve/eve-json-format.rst [hunk 2]
```diff
@@ -2915,11 +2916,14 @@ Fields
 * "cyu": List of found CYUs in the packet
 * "cyu[].hash": CYU hash
 * "cyu[].string": CYU string
+* "ja3": The JA3 fingerprint consisting of both a JA3 hash and a JA3 string
+* "ja3s": The JA3S fingerprint consisting of both a JA3 hash and a JA3 string
+* "ja4": The JA4 client fingerprint for QUIC
```
Comment (catenacyber) [inline doc/userguide/output/eve/eve-json-format.rst:2921]: What about ja4s ?

Comment (catenacyber) [reply doc/userguide/output/eve/eve-json-format.rst:2921]: Mentioned in SV PR https://github.com/OISF/suricata-verify/pull/1634/files#diff-62a9c97571df60a1b3e213d1c5951c7e32eacba95631b69e35108561a97d1d5aR35

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the work Sascha :-)

- CI : green
- Code : doing now
- Commits segmentation : Could the ja3 compile-time option be its own commit ?
- Commit messages : Could the title be `quic/tls` rather than `rust` ?
- Git ID set : looks fine for me
- CLA : you already contributed :-)
- Doc update : Ok
- Redmine ticket : ok
- Rustfmt : ok for ja4.rs
- Tests : Leaving remarks there
- Dependencies added: none

File: rust/src/lib.rs [hunk 3]
```diff
@@ -84,6 +84,8 @@ pub mod filetracker;
 pub mod kerberos;
 pub mod detect;
 
+pub mod ja4;
```
Comment (catenacyber) [inline rust/src/lib.rs:87]: Could this while file be `#[cfg(feature = "ja4")]` ?

File: rust/src/ja4.rs [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+            self.alpn[1] = char::from(alpn[alpn.len() - 1]);
+        }
+    }
+
+    pub fn add_cipher_suite(&mut self, cipher: TlsCipherSuiteID) {
+        if JA4::is_grease(u16::from(cipher)) {
+            return;
+        }
+        self.ciphersuites.push(cipher);
+        self.nof_ciphers += 1;
```
Comment (catenacyber) [inline rust/src/ja4.rs:154]: Why not use `self.ciphersuites.len()` ?

File: rust/src/ja4.rs [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+        for i in 1..200 {
+            j.add_cipher_suite(TlsCipherSuiteID(i));
+        }
+        for i in 1..200 {
+            j.add_extension(TlsExtensionType(i));
+        }
+
+        let mut s = j.get_hash();
+        s.truncate(10);
```
Comment (catenacyber) [inline rust/src/ja4.rs:322]: Could you add some comments please ?
Here we are testing the limits of cipher suites to 99, right ?

File: rust/src/quic/frames.rs [hunk 6]
```diff
@@ -284,6 +314,8 @@ fn parse_quic_handshake(msg: TlsMessage) -> Option<Frame> {
                 let mut ja3 = String::with_capacity(256);
                 ja3.push_str(&u16::from(ch.version).to_string());
                 ja3.push(',');
+                let mut ja4 = JA4::new();
```
Comment (catenacyber) [inline rust/src/quic/frames.rs:317]: Should the compile-time option prevent this creation ?

File: src/app-layer-ssl.c [hunk 7]
```diff
@@ -3048,9 +3214,10 @@ void RegisterSSLParsers(void)
         }
         SCLogDebug("ssl_config.encrypt_mode %u", ssl_config.encrypt_mode);
 
+#ifdef HAVE_JA3
+        const char *strval = NULL;
```
Comment (catenacyber) [inline src/app-layer-ssl.c:3218]: Maybe put ja3 and ja4 each in its own function guarded by the right ifdef would be more readable ?

File: doc/userguide/output/eve/eve-json-format.rst (hunk 2 above)
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:2921]: JA4 (the client fingerprint is the only one currently available as open source:
> JA4: TLS Client Fingerprinting is open-source, BSD 3-Clause, same as JA3. FoxIO does not have patent claims and is not planning to pursue patent coverage for JA4 TLS Client Fingerprinting. This allows any company or tool currently utilizing JA3 to immediately upgrade to JA4 without delay.
> JA4S, JA4L, JA4H, JA4X, JA4SSH, and all future additions, (collectively referred to as JA4+) are licensed under the FoxIO License 1.1. This license is permissive for most use cases, including for academic and internal business purposes, but is not permissive for monetization. If, for example, a company would like to use JA4+ internally to help secure their own company, that is permitted. If, for example, a vendor would like to sell JA4+ fingerprinting as part of their product offering, they would need to request an OEM license from us.

That is basically the only reason I did not implement it right now.

File: src/detect.h [hunk 8]
```diff
@@ -289,7 +289,8 @@ typedef struct DetectPort_ {
 #define SIG_FLAG_INIT_PRIO_EXPLICIT                                                                \
     BIT_U32(8) /**< priority is explicitly set by the priority keyword */
 #define SIG_FLAG_INIT_FILEDATA              BIT_U32(9)  /**< signature has filedata keyword */
-#define SIG_FLAG_INIT_JA3                   BIT_U32(10) /**< signature has ja3 keyword */
+#define SIG_FLAG_INIT_JA                    BIT_U32(10) /**< signature has ja3/ja4 keyword */
+#define SIG_FLAG_INIT_OVERFLOW              BIT_U32(11) /**< signature has overflown buffers */
```
Comment (catenacyber) [inline src/detect.h:293]: Why is this added ? Rebase leftover ?

File: src/output-json-tls.c [hunk 9]
```diff
@@ -210,6 +211,16 @@ static void JsonTlsLogJa3(JsonBuilder *js, SSLState *ssl_state)
     }
 }
 
+static void JsonTlsLogSCJA4(JsonBuilder *js, SSLState *ssl_state)
+{
+    if (ssl_state->client_connp.ja4 != NULL) {
+        uint8_t buffer[36];
+        /* JA4 hash has 36 characters */
+        SCJA4GetHash(ssl_state->client_connp.ja4, (uint8_t(*)[36])buffer);
```
Comment (catenacyber) [inline src/output-json-tls.c:219]: Do we really need the cast ?

File: doc/userguide/rules/ja4-keywords.rst (hunk 1 above)
Comment (satta) [reply doc/userguide/rules/ja4-keywords.rst:8]: Agreed. Will remove it and stick to what is implemented so far.

File: src/output-json-tls.c [hunk 10]
```diff
@@ -210,6 +211,16 @@ static void JsonTlsLogJa3(JsonBuilder *js, SSLState *ssl_state)
     }
 }
 
+static void JsonTlsLogSCJA4(JsonBuilder *js, SSLState *ssl_state)
+{
+    if (ssl_state->client_connp.ja4 != NULL) {
+        uint8_t buffer[36];
```
Comment (catenacyber) [inline src/output-json-tls.c:217]: 36 could be defined as `JA4_LEN`

File: src/util-ja3.c [hunk 11]
```diff
@@ -300,3 +302,34 @@ InspectionBuffer *Ja3DetectGetString(DetectEngineThreadCtx *det_ctx,
...
+{
+    return 0;
+}
+
+int Ja3BufferAddValue(JA3Buffer **buffer, uint32_t value)
+{
+    return 0;
+}
+
+char *Ja3GenerateHash(JA3Buffer *buffer)
```
Comment (catenacyber) [inline src/util-ja3.c:320]: Why do we need this ifndef HAVE_JA3 ?

File: src/suricata.c [hunk 12]
```diff
@@ -2687,6 +2693,10 @@ int PostConfLoadedSetup(SCInstance *suri)
 
     SetMasterExceptionPolicy();
 
+    /* Must occur prior to output mod registration
```
Comment (catenacyber) [inline src/suricata.c:2696]: Should be its own commit I guess

File: src/detect-ja4-hash.c [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+            SCLogError("JA4 support is not enabled");
+        }
+        return -2;
+    }
+    s->init_data->init_flags |= SIG_FLAG_INIT_JA;
+
+    return 0;
+}
+
+#define SC_JA4_HEX_LEN 36
```
Comment (catenacyber) [inline src/detect-ja4-hash.c:157]: Hey to be used in output ;-)

File: src/detect-ja4-hash.c [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    if (s->alproto != ALPROTO_UNKNOWN && s->alproto != ALPROTO_TLS && s->alproto != ALPROTO_QUIC) {
+        SCLogError("rule contains conflicting protocols.");
+        return -1;
+    }
+
+    /* try to enable JA4 */
+    SSLEnableJA4();
+
+    /* check if JA4 enabling had an effect */
+    if (!RunmodeIsUnittests() && !SSLJA4IsEnabled()) {
```
Comment (catenacyber) [inline src/detect-ja4-hash.c:146]: Why test `RunmodeIsUnittests` ?

File: src/detect-ja4-hash.c [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+
+#include "conf.h"
+#include "conf-yaml-loader.h"
+
+#include "util-debug.h"
+#include "util-spm.h"
+#include "util-print.h"
+#include "util-ja3.h"
+
+#include "stream-tcp.h"
```
Comment (catenacyber) [inline src/detect-ja4-hash.c:51]: Do we really need all these includes ?

File: src/app-layer-ssl.c [hunk 16]
```diff
@@ -1272,6 +1398,28 @@ static inline int TLSDecodeHSHelloExtensions(SSLState *ssl_state,
...
+                break;
+            }
+
+            case SSL_EXTENSION_ALPN: {
+                /* coverity[tainted_data] */
+                ret = TLSDecodeHSHelloExtensionALPN(ssl_state, input, ext_len);
+                if (ret < 0)
+                    goto end;
+
+                input += ret;
```
Comment (catenacyber) [inline src/app-layer-ssl.c:1418]: Why not ext_len ?

File: src/app-layer-ssl.c [hunk 17]
```diff
@@ -1171,6 +1197,106 @@ static inline int TLSDecodeHSHelloExtensionEllipticCurvePF(SSLState *ssl_state,
...
+
+            if (!(HAS_SPACE(protolen)))
+                goto invalid_length;
+
+            /* we only want the first value for JA4 */
+            if (alpn_processed_len == 1) {
+                SCJA4SetALPN(ssl_state->curr_connp->ja4, (const char *)input, protolen);
+            }
+
+            alpn_processed_len += protolen;
```
Comment (catenacyber) [inline src/app-layer-ssl.c:1283]: Still sad that this can overflow

File: doc/userguide/rules/ja4-keywords.rst (hunk 1 above)
Comment (satta) [reply doc/userguide/rules/ja4-keywords.rst:8]: Done

File: doc/userguide/output/eve/eve-json-format.rst (hunk 2 above)
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:2921]: That's just literally the output from the reference implementation tool for documentation in the S-V README (which obviously implements JA4s). I can surely remove it but since we're not implementing it I don't see why mentioning a single specific hash there should be a problem.

File: rust/src/lib.rs (hunk 3 above)
Comment (satta) [reply rust/src/lib.rs:87]: I guess we could have two modules `ja4` and `ja4stubs` and then conditionally enable either on the `lib.rs` level depending on whether the feature is set. But sure if that improves readability a lot. I can give it a try if you want.

File: rust/src/ja4.rs (hunk 4 above)
Comment (satta) [reply rust/src/ja4.rs:154]: No good reason. Will change that :)

File: rust/src/ja4.rs (hunk 5 above)
Comment (satta) [reply rust/src/ja4.rs:322]: Yep. Added a comment.

File: rust/src/quic/frames.rs (hunk 6 above)
Comment (satta) [reply rust/src/quic/frames.rs:317]: I tried not to do it that way; introducing an `Option` here would have led to more complicated code throughout the rest of the QUIC handling, which I wanted to avoid.
In case JA4 is going to be disabled the operations on that object are going to be nops.

File: src/detect.h (hunk 8 above)
Comment (satta) [reply src/detect.h:293]: Probably an effect of the rebase, yes. Not something added in the JA4 context. Since this PR has been around a lot, this is very likely a stowaway line. Removed.

File: src/output-json-tls.c (hunk 10 above)
Comment (satta) [reply src/output-json-tls.c:217]: done

File: src/output-json-tls.c (hunk 9 above)
Comment (satta) [reply src/output-json-tls.c:219]: Yes, otherwise we'd get
```
  CC       output-json-tls.o
output-json-tls.c: In function ‘JsonTlsLogSCJA4’:
output-json-tls.c:221:51: warning: passing argument 2 of ‘SCJA4GetHash’ from incompatible pointer type [-Wincompatible-pointer-types]
  221 |         SCJA4GetHash(ssl_state->client_connp.ja4, buffer);
      |                                                   ^~~~~~
      |                                                   |
      |                                                   uint8_t * {aka unsigned char *}
In file included from rust.h:24,
                 from app-layer-events.h:30,
                 from app-layer-parser.h:28,
                 from output-json-tls.c:40:
./../rust/gen/rust-bindings.h:5023:46: note: expected ‘uint8_t (*)[36]’ {aka ‘unsigned char (*)[36]’} but argument is of type ‘uint8_t *’ {aka ‘unsigned char *’}

```
I guess we could remove the string length from the interface of `SCJA4GetHash` but I'd rather have that checked.

File: src/detect-tls-ja3s-string.c [hunk 18]
```diff
@@ -56,6 +56,27 @@
 #include "util-unittest.h"
 #include "util-unittest-helper.h"
 
+#ifndef HAVE_JA3
+
+static int DetectJA3SetupNoSupport(DetectEngineCtx *a, Signature *b, const char *c)
+{
+    SCLogError("no JA3 support built in");
```
[the remark this answers is not in the document]
Comment (satta) [reply src/detect-tls-ja3s-string.c:63]: Yes, properly using `requires` would just cause the rule containing such a keyword to be skipped. That's what the `*-requires-off` S-V tests are meant to check.

File: src/detect-ja4-hash.c (hunk 13 above)
Comment (satta) [reply src/detect-ja4-hash.c:157]: I moved this into a separate header to be used in output and detect.

File: src/detect-ja4-hash.c (hunk 15 above)
Comment (satta) [reply src/detect-ja4-hash.c:51]: Absolutely not. Total artifact of code reuse. I just reduced this to less than half.

File: src/util-ja3.c (hunk 11 above)
Comment (satta) [reply src/util-ja3.c:320]: I wanted to keep the behaviour as close as possible to the `HAVE_JA3` case. But I guess we can also pass `NULL` here, given that it won't be accessed in that case and also will only be freed if it is non-`NULL`.

File: src/detect-ja4-hash.c (hunk 14 above)
Comment (satta) [reply src/detect-ja4-hash.c:146]: Inspired by JA3 code; I figured that for unittests it should not matter if something is enabled or disabled in the configuration.

File: src/app-layer-ssl.c (hunk 17 above)
Comment (satta) [reply src/app-layer-ssl.c:1283]: Why?
* `alpn_len` is a 16-bit unsigned value.
* `alpn_processed_len` is a 32-bit unsigned value.
* `protolen` is a 8-bit unsigned value, i.e. `alpn_processed_len` can only be incremented each iteration by at most 255.
* So the only way that this can overflow is that `alpn_processed_len` + `protolen` exceeds the range of a 32-bit unsigned value -- but that cannot happen since in that case `alpn_processed_len` would have had to have exceeded any value of `alpn_len` many iterations ago, since we only can increment `alpn_processed_len` by 255 at most. But that would have already been caught by the `while (alpn_processed_len < alpn_len)` comparison at the largest 16-bit unsigned value at the latest, so we couldn't even get into that region.

What am I missing?

File: src/suricata.c (hunk 12 above)
Comment (satta) [reply src/suricata.c:2696]: Yeah, makes sense I guess.

File: src/app-layer-ssl.c (hunk 7 above)
Comment (satta) [reply src/app-layer-ssl.c:3218]: Can do.

File: rust/src/lib.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/lib.rs:87]: Oh, so I think my first question should be :
Why should we need `ja4stubs` if ja4 is not enabled ?

File: src/output-json-tls.c (hunk 9 above)
Comment (catenacyber) [reply src/output-json-tls.c:219]: This is strange from the compiler...

File: src/detect-ja4-hash.c (hunk 14 above)
Comment (catenacyber) [reply src/detect-ja4-hash.c:146]: fair enough even if it looks still mysterious to me

File: src/app-layer-ssl.c (hunk 17 above)
Comment (catenacyber) [reply src/app-layer-ssl.c:1283]: Oh yes,

> alpn_processed_len is a 32-bit unsigned value 

This is good

But still, I can have input going further than `alpn_len` if the last `protolen` is too big, and so we may parse bad wrongly next extensions...

File: src/util-ja3.c (hunk 11 above)
Comment (catenacyber) [reply src/util-ja3.c:320]: But why should we need an empty `Ja3GenerateHash` if ja3 is not enabled ?

File: rust/src/quic/frames.rs (hunk 6 above)
Comment (catenacyber) [reply rust/src/quic/frames.rs:317]: I was not talking about `Option` but `#cfg[ja4]` (or something like that)

**Comment by catenacyber** [comment]
Thanks Sascha, waiting for the next version

File: rust/src/lib.rs (hunk 3 above)
Comment (satta) [reply rust/src/lib.rs:87]: Because we need to have the Rust code provide symbols for the `SCJA4*` functions. otherwise we'd need to conditionally remove all the calls to JA4 state modifying code in the QUIC and TLS parsers. Yes, for the Rust code there is an `Option`, but for the C side we would otherwise need `#ifdef`s all over the place if the Rust module is not compiled. I didn't want to clutter the parser libraries with the details of whether JA4 is enabled or not, but let the implementation decide whether to actually do something or not. Hence the use of stubs that do not do anything when JA4 is disabled.

File: src/output-json-tls.c (hunk 9 above)
Comment (satta) [reply src/output-json-tls.c:219]: That's gcc 12.2.0  🤷🏻  I'd just keep the cast in. I will replace the hardcoded length with a preprocessor constant.

File: src/util-ja3.c (hunk 11 above)
Comment (satta) [reply src/util-ja3.c:320]: See above - I didn't want to have `#ifdefs` in the parser code and just base the behaviour on the implementation of these functions.

File: src/app-layer-ssl.c (hunk 17 above)
Comment (satta) [reply src/app-layer-ssl.c:1283]: Ah, so your comment was not about an integer _type_ overflow (that might cause a loop), but rather a semantic error to be able to exceed `alpn_len`...

**Comment by satta** [comment]
Will take a closer look while on a train ride today. Sorry for the slow responses here but I'm rather busy at other fronts arm.

File: doc/userguide/output/eve/eve-json-format.rst (hunk 2 above)
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:2921]: Just removed from S-V anyway :)

**Comment by victorjulien** [comment]
> Will take a closer look while on a train ride today. Sorry for the slow responses here but I'm rather busy at other fronts atm.

No worries, we appreciate the effort!

File: src/suricata.c (hunk 12 above)
Comment (satta) [reply src/suricata.c:2696]: done

File: src/app-layer-ssl.c (hunk 17 above)
Comment (catenacyber) [reply src/app-layer-ssl.c:1283]: My comment was more generically speaking about a missing check and all that could go wrong because of it ;-)

File: rust/src/lib.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/lib.rs:87]: > otherwise we'd need to conditionally remove all the calls to JA4 state modifying code in the QUIC and TLS parsers

Yes, that is what I would have done.

But I see your point about let the JA4 implementation decide whether to actually do something or not 👍

File: src/app-layer-ssl.c (hunk 17 above)
Comment (satta) [reply src/app-layer-ssl.c:1283]: So I guess it would be safer to do something like that, to recheck whether it's OK to read from the buffer and increment `input`?
```diff
diff --git a/src/app-layer-ssl.c b/src/app-layer-ssl.c
index 361a81696..16cc2bc66 100644
--- a/src/app-layer-ssl.c
+++ b/src/app-layer-ssl.c
@@ -1275,7 +1275,14 @@ static inline int TLSDecodeHSHelloExtensionALPN(
             if (!(HAS_SPACE(protolen)))
                 goto invalid_length;
 
-            /* we only want the first value for JA4 */
+            /* Check if this length would exceed the overall ALPN length;
+               if so, skip and continue */
+            if (alpn_processed_len + protolen > ((uint32_t) alpn_len)) {
+                input += alpn_len - alpn_processed_len;
+                break;
+            }
+
+            /* Only record the first value for JA4 */
             if (alpn_processed_len == 1) {
                 SCJA4SetALPN(ssl_state->curr_connp->ja4, (const char *)input, protolen);
             }

```

**Comment by satta** [comment]
Next: #10579 10579
